rules/group.go

```diff
@@ -542,7 +554,12 @@ func (g *Group) Eval(ctx context.Context, ts time.Time) {
 			rule.SetHealth(HealthBad)
 			rule.SetLastError(err)
 			sp.SetStatus(codes.Error, err.Error())
-			g.metrics.EvalFailures.WithLabelValues(GroupKey(g.File(), g.Name())).Inc()
+
+			if _, ok := rule.(*AlertingRule); ok {
+				g.metrics.EvalFailures.WithLabelValues(GroupKey(g.File(), g.Name()), KindAlerting).Inc()
+			} else {
+				g.metrics.EvalFailures.WithLabelValues(GroupKey(g.File(), g.Name()), KindRecording).Inc()
```

Comment:
This kind of type assertion is ugly, and fails silently if someone is using the code with a different type.
Suggest to add a `Kind()` method to the `Rule` interface.

Reply:
(Note that the API already builds and exposes a "type" https://github.com/prometheus/prometheus/blob/2ef8706c275bf06dcf482605abcd4def9e2d3af5/web/api/v1/api.go#L1454 maybe we could move that here as Bryan suggested and get rid of all the other assertions)

Reply:
The updated code still "fails silently if someone is using the code with a different type".

"Ugly" is subjective but I still feel that way.

> could also use the `Type` from api.go

I think the suggestion is to move the implementation from `api` to `rules`, which is a bit more upheaval, but would remove the switch statements.